ethereum/core/src/main/java/org/hyperledger/besu/ethereum/transaction/PreWorldStateCloseGuard.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.transaction;
+
+import org.hyperledger.besu.ethereum.core.MutableWorldState;
+
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+/**
+ * This class is passed to the transaction simulator and allows executing code before the copy of
+ * the worldstate is closed. When processing a transaction with the simulator, a copy of the
+ * worldstate is created, and once the simulation process is complete, the copy is closed. In some
+ * cases the simulator fills a Tracer that requires updates from the worldstate. If we want to
+ * generate a trace after the execution, it would fail. This interface allows us to execute our code
+ * before the worldstate is closed, ensuring that we can generate our objects without any issues.
+ * Then, the simulator will properly close the copy of the worldstate.
+ */
+public interface PreWorldStateCloseGuard<U>
```

Comment:
IIUC, would `onWorldStateClose` make sense?

Reply:
not sure this interface is not closed when the worldstate is close but before

Reply:
hmm yes I see how you got to your name, tricky one.

`worldStateWillClose()` 🤷 😆 

nit: PreWorldStateCloseGuard function name is a noun rather than a verb

Reply:
I will try to think better for another pr

Reply:
I will use that i think ilt will be better PreCloseStateHandler